[SE-0163] Migrate from deprecated Unicode APIs

benrimmington opened the pull request:
`UnicodeCodec`, `UnicodeDecodingResult`, and some typealiases were deprecated by the [SE-0163][] proposal.

[SE-0163]: <https://github.com/apple/swift-evolution/blob/master/proposals/0163-string-revision-1.md>

benrimmington commented on benchmark/single-source/UTF8Decode.swift:
```diff
@@ -115,7 +115,7 @@ public func run_UTF8Decode(_ N: Int) {
     for string in strings {
       var it = string.makeIterator()
-      var utf8 = UTF8()
+      var utf8 = Unicode.UTF8()
       while !isEmpty(utf8.decode(&it)) { }
```
This benchmark is using deprecated APIs.

benrimmington replied:
This benchmark is using deprecated APIs.

benrimmington commented on stdlib/public/core/MigrationSupport.swift:
```diff
@@ -2,16 +2,23 @@
+//
+// Placeholders:
+//
+// * `deprecated: 100000` is API_TO_BE_DEPRECATED in the next version of Swift.
+// * `obsoleted: TODO` is for a future (source-breaking) language version mode.
```
This is for "soft deprecation" (until we know the next Swift version number).

See also: [swift-corelibs-foundation](https://github.com/apple/swift-corelibs-foundation/blob/514c9e9ea90bbb6596bfbf343db205d4d588cc56/Sources/Foundation/FileHandle.swift#L646-L694)

benrimmington commented on stdlib/public/core/StaticString.swift:
```diff
@@ -182,7 +182,7 @@ public struct StaticString {
         i += 1
       }
-      UTF8.encode(unicodeScalar, into: sink)
+      Unicode.UTF8.encode(unicodeScalar, into: sink)
```
This `@_transparent` StaticString method is using deprecated APIs.

benrimmington replied:
The internal [`Unicode.Scalar.withUTF8CodeUnits(_:)`][1] method has been ABI-public since Swift 5.0, so I tried changing the else branch to:

```swift
return unicodeScalar.withUTF8CodeUnits { body($0) }
```

There was a failure in [test/IRGen/objc_super.swift:139:11][2]:

```
error: CHECK: expected string not found in input
// CHECK: @"$ss12StaticStringV14withUTF8BufferyxxSRys5UInt8VGXElFyAEcfU_"
          ^
<stdin>:2290:1: note: scanning from here
entry:
^
<stdin>:2308:41: note: possible intended match here
define linkonce_odr hidden swiftcc void @"$ss12StaticStringV14withUTF8BufferyxxSRys5UInt8VGXElFxAFXEfU_"(%swift.opaque* noalias nocapture sret %0, i64 %1, i64 %2, i8* %3, %swift.opaque* %4, %swift.type* %R) #0 {
```

I think the expected mangled name contains the [`let sink: (UInt8) -> Void`][3] closure of the existing implementation.

[1]: <https://github.com/apple/swift/blob/swift-5.0-branch/stdlib/public/core/UnicodeScalar.swift#L470-L482>
[2]: <https://github.com/apple/swift/blob/3e090b483352b3c6f889eec521605fda3ec1ab51/test/IRGen/objc_super.swift#L139>
[3]: <https://github.com/apple/swift/blob/3e090b483352b3c6f889eec521605fda3ec1ab51/stdlib/public/core/StaticString.swift#L177>

benrimmington commented on stdlib/public/core/Unicode.swift:
```diff
@@ -30,23 +31,6 @@ public enum UnicodeDecodingResult: Equatable {
   /// An indication of a decoding error.
   case error
-
-  @inlinable
-  public static func == (
-    lhs: UnicodeDecodingResult,
-    rhs: UnicodeDecodingResult
-  ) -> Bool {
-    switch (lhs, rhs) {
-    case (.scalarValue(let lhsScalar), .scalarValue(let rhsScalar)):
-      return lhsScalar == rhsScalar
-    case (.emptyInput, .emptyInput):
-      return true
-    case (.error, .error):
-      return true
-    default:
-      return false
-    }
-  }
```
This can be synthesized by the compiler.

benrimmington commented on stdlib/public/core/Unicode.swift:
```diff
@@ -276,6 +262,9 @@ extension Unicode.UTF8: UnicodeCodec {
     if _fastPath(s == 0) { return }
     processCodeUnit(UInt8(truncatingIfNeeded: s) &- 0x01)
   }
+}
+
+extension Unicode.UTF8 {
```
The non-deprecated APIs in this extension could be moved into UTF8.swift:

* `Unicode.UTF8.isContinuation(_:)`
* `Unicode.UTF8._nullCodeUnitOffset(in:)`

benrimmington replied:
The _non-deprecated_ APIs could be moved into UTF8.swift:
* `Unicode.UTF8.isContinuation(_:)`
* `Unicode.UTF8._nullCodeUnitOffset(in:)`
* `Unicode.UTF8._decodeOne(_:)` (used in [UnicodeLongTest.swift][])

The _deprecated_ APIs could be moved into MigrationSupport.swift:
* `UnicodeDecodingResult` enum,
* `UnicodeCodec` protocol and conformances.

The _unused_ APIs could be moved into LegacyABI.swift:
* `Unicode.UTF16._decodeOne(_:)`
* `transcode(_:_:_:_:stopOnError:)` ([unavailable][] since Swift 3).

[unavailable]: <https://github.com/apple/swift/blob/swift-3.0-branch/stdlib/public/core/Unicode.swift#L1152-L1164>
[UnicodeLongTest.swift]: <https://github.com/apple/swift/blob/master/validation-test/stdlib/UnicodeLongTest.swift>

benrimmington commented on validation-test/stdlib/UnicodeUTFEncoders.swift:
```diff
@@ -14,13 +14,18 @@ import StdlibUnittest
 
-protocol TestableUnicodeCodec : UnicodeCodec {
-  associatedtype CodeUnit : FixedWidthInteger
+protocol TestableUnicodeCodec: Unicode.Encoding {
   static func encodingId() -> String.Encoding
   static func name() -> NSString
+
+  // [SE-0163] Test the deprecated `UnicodeCodec` APIs.
+  static func encode(
+    _ input: Unicode.Scalar,
+    into processCodeUnit: (CodeUnit) -> Void
+  )
```
This should allow deprecated APIs to be tested, without using the `UnicodeCodec` protocol.

benrimmington reviewed (commented):
See also: #33167, #33175

benrimmington commented:
Should the deprecated APIs be moved?

* Unicode.swift → LegacyABI.swift
  - `Unicode.UTF8._decodeOne(_:)`
  - `Unicode.UTF16._decodeOne(_:)`
  - `transcode(_:_:_:_:stopOnError:)` ([unavailable][] since Swift 3).
* Unicode.swift → MigrationSupport.swift
  - `UnicodeDecodingResult` enum,
  - `UnicodeCodec` protocol and conformances.

[unavailable]: <https://github.com/apple/swift/blob/swift-3.0-branch/stdlib/public/core/Unicode.swift#L1152-L1164>

benrimmington commented:
Cc: @airspeedswift, @milseman

benrimmington commented on stdlib/public/core/StaticString.swift:
```diff
@@ -172,20 +172,7 @@ public struct StaticString {
-      return body(UnsafeBufferPointer(
-        start: UnsafePointer(Builtin.addressof(&buffer)),
-        count: i))
+      return unicodeScalar.withUTF8CodeUnits { body($0) }
```
This is using the internal [`Unicode.Scalar.withUTF8CodeUnits(_:)`][1] method, which has been ABI-public since Swift 5.0 — I assume it will also be available in the bundled stdlib (e.g. when targeting macOS 10.9).

[1]: <https://github.com/apple/swift/blob/swift-5.0-branch/stdlib/public/core/UnicodeScalar.swift#L470-L482>

benrimmington replied:
This is using the internal [`Unicode.Scalar.withUTF8CodeUnits(_:)`][1] method, which has been ABI-public since Swift 5.0 — I assume it will also be available in the bundled stdlib (e.g. when targeting macOS 10.9).

[1]: <https://github.com/apple/swift/blob/swift-5.0-branch/stdlib/public/core/UnicodeScalar.swift#L470-L482>

benrimmington commented on test/IRGen/objc_super.swift:
```diff
@@ -136,5 +136,5 @@ class GenericRuncer<T> : Gizmo {
 
 // CHECK: define internal swiftcc void [[PARTIAL_FORWARDING_THUNK]](%swift.refcounted* swiftself %0) {{.*}} {
-// CHECK: @"$ss12StaticStringV14withUTF8BufferyxxSRys5UInt8VGXElFyAEcfU_"
+// CHECK: @"$ss12StaticStringV14withUTF8BufferyxxSRys5UInt8VGXElFxAFXEfU_"
```
This test now has a different `closure #1` type in its mangled name.

* Old swift-demangle result:

  ```
  $ss12StaticStringV14withUTF8BufferyxxSRys5UInt8VGXElFyAEcfU_ ---> 
  closure #1 (Swift.UInt8) -> () in 
  Swift.StaticString.withUTF8Buffer<A>((Swift.UnsafeBufferPointer<Swift.UInt8>) -> A) -> A
  ```

* New swift-demangle result:

  ```
  $ss12StaticStringV14withUTF8BufferyxxSRys5UInt8VGXElFxAFXEfU_ ---> 
  closure #1 (Swift.UnsafeBufferPointer<Swift.UInt8>) -> A in 
  Swift.StaticString.withUTF8Buffer<A>((Swift.UnsafeBufferPointer<Swift.UInt8>) -> A) -> A
  ```

benrimmington replied:
This test now has a different `closure #1` type in its mangled name.

* Old swift-demangle result:

  ```
  $ss12StaticStringV14withUTF8BufferyxxSRys5UInt8VGXElFyAEcfU_ ---> 
  closure #1 (Swift.UInt8) -> () in 
  Swift.StaticString.withUTF8Buffer<A>((Swift.UnsafeBufferPointer<Swift.UInt8>) -> A) -> A
  ```

* New swift-demangle result:

  ```
  $ss12StaticStringV14withUTF8BufferyxxSRys5UInt8VGXElFxAFXEfU_ ---> 
  closure #1 (Swift.UnsafeBufferPointer<Swift.UInt8>) -> A in 
  Swift.StaticString.withUTF8Buffer<A>((Swift.UnsafeBufferPointer<Swift.UInt8>) -> A) -> A
  ```

benrimmington commented on validation-test/stdlib/UnicodeUTFEncoders.swift:
```diff
@@ -14,13 +14,12 @@ import StdlibUnittest
 
-protocol TestableUnicodeCodec : UnicodeCodec {
-  associatedtype CodeUnit : FixedWidthInteger
+protocol TestableUnicodeCodec: Unicode.Encoding {
```
This file has been updated to use only non-deprecated APIs.

I should probably keep a copy of the old file, to run in compatibility mode.

benrimmington commented on validation-test/stdlib/UnicodeUTFEncoders.swift:
```diff
@@ -1,7 +1,5 @@
-// RUN: %target-build-swift %s -o %t/a.out -O
-// RUN: %target-codesign %t/a.out
-// RUN: %target-run %t/a.out
+// RUN: %target-run-simple-swift(-O -swift-version 4)
+// RUN: %target-run-simple-swift(-O -swift-version 5)
```
Deprecated APIs are only tested in Swift 4 compatibility mode.

benrimmington replied:
The deprecated APIs are only tested in Swift 4 compatibility mode.

benrimmington commented:
@swift-ci Please smoke test

benrimmington commented on stdlib/public/core/MigrationSupport.swift:
```diff
@@ -2,16 +2,23 @@
-// of Swift
+// of Swift.
+//
+// Placeholders:
+//
+// * `deprecated: 100000` is API_TO_BE_DEPRECATED in the next version of Swift.
+// * `obsoleted: TODO` is for a future (source-breaking) language version mode.
+//
+//===----------------------------------------------------------------------===//
```
This is for "soft deprecation" (until we know the next Swift version number).

See also: [swift-corelibs-foundation](https://github.com/apple/swift-corelibs-foundation/blob/514c9e9ea90bbb6596bfbf343db205d4d588cc56/Sources/Foundation/FileHandle.swift#L646-L694)

benrimmington commented on stdlib/public/core/Unicode.swift:
```diff
@@ -30,23 +31,6 @@ public enum UnicodeDecodingResult: Equatable {
   /// An indication of a decoding error.
   case error
-
-  @inlinable
-  public static func == (
-    lhs: UnicodeDecodingResult,
-    rhs: UnicodeDecodingResult
-  ) -> Bool {
-    switch (lhs, rhs) {
-    case (.scalarValue(let lhsScalar), .scalarValue(let rhsScalar)):
-      return lhsScalar == rhsScalar
-    case (.emptyInput, .emptyInput):
-      return true
-    case (.error, .error):
-      return true
-    default:
-      return false
-    }
-  }
```
This can be synthesized by the compiler.

lorentey replied:
This is not exactly the same, because the synthesized definition isn’t inlinable. How that will affect performance/code size of code using these APIs is difficult to predict.

benrimmington commented:
@CodaFi Please could you review, and merge if possible?

I've added some comments above, for anything which requires more attention or follow-up PRs.

CodaFi reviewed (approved):
Benchmarks and test changes LGTM. But @milseman or @airspeedswift should sign off on the stdlib changes

CodaFi commented:
@swift-ci benchmark

milseman commented:
I realize this went through SE quite a while ago, but I think it's a bad idea to formally remove the Unicode aliases, especially for UTF8/UTF16.

benrimmington commented:
@milseman Can we keep the typealiases at `deprecated: 100000` indefinitely?

I've copied the simple renames into #33167, in case they can be merged sooner; and to make it easier to review this PR (after rebasing).

lorentey commented:
> Can we keep the typealiases at `deprecated: 100000` indefinitely?

Why? The version number 100000 has no special meaning, and it should not appear in any compiler diagnostics or generated documentation. If the reason why SE-0163 intended to deprecate these names still stands, then I think we should reconfirm this through swift-evolution and set the deprecation to a valid version number.

benrimmington commented:
@lorentey Version number 100000 is from the `API_TO_BE_DEPRECATED` macro in the `<os/availability.h>` header.

It seems to affect only the documentation, at least when applied to an OS availability attribute.

For example, in SwiftUI the [`colorScheme(_:)` method][1] has the generated interface:

```swift
@available(iOS 13.0, OSX 10.15, tvOS 13.0, watchOS 6.0, *)
extension View {

    /// Sets this view's color scheme.
    ///
    /// Use `colorScheme(_:)` to set the color scheme for the view to which you
    /// apply it and any subviews. If you want to set the color scheme for all
    /// views in the presentation, use ``View/preferredColorScheme(_:)``
    /// instead.
    ///
    /// - Parameter colorScheme: The color scheme for this view.
    ///
    /// - Returns: A view that sets this view's color scheme.
    @available(iOS, introduced: 13.0, deprecated: 100000.0, renamed: "preferredColorScheme(_:)")
    @available(OSX, introduced: 10.15, deprecated: 100000.0, renamed: "preferredColorScheme(_:)")
    @available(tvOS, introduced: 13.0, deprecated: 100000.0, renamed: "preferredColorScheme(_:)")
    @available(watchOS, introduced: 6.0, deprecated: 100000.0, renamed: "preferredColorScheme(_:)")
    @inlinable public func colorScheme(_ colorScheme: ColorScheme) -> some View
}
```

[1]: <https://developer.apple.com/documentation/swiftui/view/colorscheme(_:)>

benrimmington committed:
[SE-0163] Migrate from deprecated Unicode APIs

benrimmington commented:
@swift-ci Please smoke test

lorentey commented:
> Version number 100000 is from the API_TO_BE_DEPRECATED macro in the <os/availability.h> header.

Unfortunately AFAIK that only applies to C. :-( We could adopt it in Swift, but without the macro name the number seems opaque. 

The 9999s we use elsewhere set an unfortunate precedent for doing such things; I guess it's time to replace them too with a better system.

milseman commented on stdlib/public/core/MigrationSupport.swift:
```diff
@@ -237,33 +244,31 @@ extension String {
-//===--- Slicing Support --------------------------------------------------===//
-
-// @available(swift,deprecated: 5.0, renamed: "Unicode.UTF8")
+@available(swift, deprecated: 100000, /*obsoleted: TODO,*/ renamed: "Unicode.UTF8")
```
Also, that batch of SEs were part of a larger attempt to reformulate how strings were formulated that didn't pan out in practice. They have several portions that weren't fully thought through, much less implemented, so I don't think we want to blindly apply them without some reevaluation.
